### ISSUE TITLE
A question about the beginning of the GRU

### ISSUE THREAD

**Issue by menglin0320** [JianshuZhang/WAP#2]
Maybe there is a convention on the field but I don't know, so at the beginning there is no h_{-1}, how do we compute c_{0}? or simply to get h0, what is h_{-1},a0,f0? I feel you didn't mention this on the paper. Maybe it's too trivial?

**Comment by menglin0320** [comment]
Found the answer on the attention mechanism paper. Close the comment, sorry.

**Comment by JianshuZhang** [comment]
It's OK. Feel free to give a comment if you have questions.

**Comment by menglin0320** [comment]
well I'm not famaliar with Theano, I just wonder where can I see the model?
There are three questions about some details of your model.
The paper Neural Image Caption generator with Visual Attention stated that 
**1. The  initial memory state and hidden stated of the LSTM are predicted by an average of the annotation vectors fed through two separate MLPS. So for those MLPs, what is your choice?**
**2. And Can I know the kernel size of the conv net that transforms beta_{t} to F**
**3. And here $sum_{l}^{t-1} \alpha_l$ do you mean $sum_{l=0}^{t-1} \alpha_l $**
Sorry I don't know how to write the  Inline Rendering code for github comments.

 
![image](https://user-images.githubusercontent.com/14263971/37577428-22c509ea-2b00-11e8-9a4a-567ee805660a.png)

**Comment by JianshuZhang** [comment]
The models stored in .npz in DenseNet is well pre-trained, you can see the model by import numpy.
As for your question:
1. As I employed GRU, there is no memory state. The hidden state is predicted by average of annotation vectors fed through MLP and the MLP is random initialized.
2. convolutional kernel size is 11 x 11
3. Yes, it means sum_{l=0}^{t-1}\alpha_l



xysszjs@mail.ustc.edu.cn
 
From: menglin0320
Date: 2018-03-19 11:53
To: JianshuZhang/WAP
CC: Jianshu Zhang; Comment
Subject: Re: [JianshuZhang/WAP] A question about the beginning of the GRU (#2)
well I'm not famaliar with Theano, I just wonder where can I see the model?
There are three questions about some details of your model.
The paper Neural Image Caption generator with Visual Attention stated that
1. The initial memory state and hidden stated of the LSTM are predicted by an average of the annotation vectors fed through two separate MLPS. So for those MLPs, what is your choice?
2. And Can I know the kernel size of the conv net that transforms beta_{t} to F
3. And here $sum_{l}^{t-1} \alpha_l$ do you mean $sum_{l=0}^{t-1} \alpha_l $
Sorry I don't know how to write the Inline Rendering code for github comments.

**Comment by menglin0320** [comment]
Thanks， and since I'm new to attention mechanism, this question may be basic. For this line
![image](https://user-images.githubusercontent.com/14263971/37579893-dae38c50-2b0e-11e8-84d8-ad416b4c0fd7.png)
1. for example if the feature map(encoded latent) has shape(14,14,512)
There will be 512 different Wa, Ua, and Va, instead of just one Wa, Ua, and Va. I'm pretty sure my guessing is true since only one Wa doesn't make sense at all, but I want to make sure.

2.Well I wonder if it is ok to know how you defined the MLP for h0 and c0... if the paper didn't mention MLP I would simply use one fully connected layer.. I wonder if the convention is to use standard networks like 3 layers of (3*3 kernel convolution with relu and batch norm)?

And I wonder if there is a way to see how you define the model? I tried to load the npy but it's only bounch of parameters,(though the keys provides some information). I just want to know the structure of the model.
Thanks for your patience.

**Comment by JianshuZhang** [comment]
Well, there is **ONLY ONE** Wa, Ua and Va. If the feature map has shape (14,14,512), then there will be 14x14 different a_i, each a_i has 512 dimension. At each time step t and each pixel i, h_t-1 and a_i are different, which lead to different e_ti.
The MLP is one fully connected layer. As for the convention, I guess you mean convolution? The details of convolution architecture can be seen in my paper, relu and batch norm are both needed.
Unlike  Keras or some other tools, Theano CANNOT show the structure of the model. The details of the structure of the model can be seen in my paper, in the first paper (Watch, Attend and Parse) only the implementation of coverage model is ignored.

**Comment by menglin0320** [comment]
An artical said MLP means at least three layers network so I was confused. You've answered my question, one fully connected layer is enough, sorry for confuse you.

Yes I misunderstood. And what made me doubt that there are not only one Wa is the fact that h_{t-1} is the same for all i at timestep t, W_a h_{t-1} only shifts the whole thing. But it makes sense because this shift changes the softmax output(alpha_t).

And one last question what did you pick for attention dimension? the n prime.

**Comment by JianshuZhang** [comment]
In my experiments, I choose the attention dimension as the annotation dimension. For instance, if the feature map has shape (14,14,512), then the attention dimension is 512. In the released codes, the attention dimension in 'VGG' is 128 while the attention dimension in 'DenseNet' is 512.

**Comment by menglin0320** [comment]
Thankyou！

**Comment by menglin0320** [comment]
Hmmm, just a quick question, if everything works fine, we should see a big drop on the first few iterations or not.

**Comment by JianshuZhang** [comment]
WER and ExpRate metric of testing set begin to get better after about 10 epochs. Training cost drops a lot on the first few iterations.
